network/p2p/peer_tracker.go

```diff
@@ -137,150 +145,166 @@ func (p *PeerTracker) shouldTrackNewPeer() bool {
+
+	if nodeID, ok := p.trackedPeers.Peek(); ok {
+		p.log.Debug("selecting peer",
+			zap.String("reason", "tracked"),
+			zap.Stringer("nodeID", nodeID),
+			zap.Bool("checkedBandwidthHeap", useBandwidthHeap),
+		)
+		return nodeID, true
 	}
-	p.log.Debug(
-		"peer tracking: popping peer",
-		zap.Stringer("nodeID", nodeID),
-		zap.Bool("random", useRand),
-	)
-	return nodeID, true
+
+	// We're not connected to any peers.
+	return ids.EmptyNodeID, false
 }
 
 // Record that we sent a request to [nodeID].
-func (p *PeerTracker) TrackPeer(nodeID ids.NodeID) {
+//
+// Removes the peer's bandwidth averager from the bandwidth heap.
+func (p *PeerTracker) RegisterRequest(nodeID ids.NodeID) {
 	p.lock.Lock()
 	defer p.lock.Unlock()
 
+	p.untrackedPeers.Remove(nodeID)
 	p.trackedPeers.Add(nodeID)
+	p.bandwidthHeap.Remove(nodeID)
+
 	p.numTrackedPeers.Set(float64(p.trackedPeers.Len()))
 }
 
 // Record that we observed that [nodeID]'s bandwidth is [bandwidth].
+//
 // Adds the peer's bandwidth averager to the bandwidth heap.
-func (p *PeerTracker) TrackBandwidth(nodeID ids.NodeID, bandwidth float64) {
+func (p *PeerTracker) RegisterResponse(nodeID ids.NodeID, bandwidth float64) {
+	p.updateBandwidth(nodeID, bandwidth, true)
+}
+
+// Record that a request failed to [nodeID].
+//
+// Adds the peer's bandwidth averager to the bandwidth heap.
+func (p *PeerTracker) RegisterFailure(nodeID ids.NodeID) {
+	p.updateBandwidth(nodeID, 0, false)
+}
+
+func (p *PeerTracker) updateBandwidth(nodeID ids.NodeID, bandwidth float64, responsive bool) {
 	p.lock.Lock()
 	defer p.lock.Unlock()
 
-	peer := p.peers[nodeID]
-	if peer == nil {
-		// we're not connected to this peer, nothing to do here
-		p.log.Debug("tracking bandwidth for untracked peer", zap.Stringer("nodeID", nodeID))
+	if !p.trackedPeers.Contains(nodeID) {
+		// we're not tracking this peer, nothing to do here
+		p.log.Debug("tracking bandwidth for untracked peer",
+			zap.Stringer("nodeID", nodeID),
+		)
 		return
 	}
 
 	now := time.Now()
-	if peer.bandwidth == nil {
-		peer.bandwidth = safemath.NewAverager(bandwidth, bandwidthHalflife, now)
+	peerBandwidth, ok := p.peerBandwidth[nodeID]
+	if ok {
+		peerBandwidth.Observe(bandwidth, now)
 	} else {
-		peer.bandwidth.Observe(bandwidth, now)
+		peerBandwidth = safemath.NewAverager(bandwidth, bandwidthHalflife, now)
+		p.peerBandwidth[nodeID] = peerBandwidth
 	}
-	p.bandwidthHeap.Push(nodeID, peer.bandwidth)
+	p.bandwidthHeap.Push(nodeID, peerBandwidth)
+	p.averageBandwidth.Observe(bandwidth, now)
 
-	if bandwidth == 0 {
-		p.responsivePeers.Remove(nodeID)
-	} else {
+	if responsive {
 		p.responsivePeers.Add(nodeID)
-		// TODO danlaine: shouldn't we add the observation of 0
-		// to the average bandwidth in the if statement?
-		p.averageBandwidth.Observe(bandwidth, now)
-		p.averageBandwidthMetric.Set(p.averageBandwidth.Read())
+	} else {
+		p.responsivePeers.Remove(nodeID)
 	}
+
 	p.numResponsivePeers.Set(float64(p.responsivePeers.Len()))
```

Comment:
The reason `averageBandwidthMetric` is suffixed with `Metric` was to resolve the conflict with `averageBandwidth`. I moved all the metrics into their own `metrics` struct so that we could have our own separated namespace.